README.md
<h1>:alarm_clock: timetrace
<a href="https://circleci.com/gh/dominikbraun/timetrace"><img src="https://circleci.com/gh/dominikbraun/timetrace.svg?style=shield"></a>
<a href="https://www.codefactor.io/repository/github/dominikbraun/timetrace"><img src="https://www.codefactor.io/repository/github/dominikbraun/timetrace/badge" /></a>
<a href="https://github.com/dominikbraun/timetrace/releases"><img src="https://img.shields.io/github/v/release/dominikbraun/timetrace?sort=semver"></a>
<a href="LICENSE"><img src="https://img.shields.io/badge/license-Apache--2.0-brightgreen"></a>
</h1>

> timetrace is a simple CLI for tracking your working time.

![CLI screenshot 64x16](timetrace.png)

:fire: **New:** [Add tags for records](#start-tracking)  
:fire: **New:** [Use decimal hours when displaying durations](#prefer-decimal-hours-for-status-and-reports)  
:fire: **New:** [Restore records when restoring the associated project](#delete-a-project)  
:fire: **New:** [Support for per-project configuration](#per-project-configuration)  

---

- [Installation](#installation)
  - [Homebrew](#homebrew)
  - [Snap](#snap)
  - [AUR](#aur)
  - [Scoop](#scoop)
  - [Docker](#docker)
  - [Binary](#binary)
- [Usage example](#usage-example)
  - [Project modules](#project-modules)
- [Shell integration](#shell-integration)
  - [Starship](#starship)
- [Command reference](#command-reference)
  - [Start tracking](#start-tracking)
  - [Print the tracking status](#print-the-tracking-status)
  - [Stop tracking](#stop-tracking)
  - [Create a project](#create-a-project)
  - [Create a record](#create-a-record)
  - [Get a project](#get-a-project)
  - [Get a record](#get-a-record)
  - [List all projects](#list-all-projects)
  - [List all records from a date](#list-all-records-from-a-date)
  - [Edit a project](#edit-a-project)
  - [Edit a record](#edit-a-record)
  - [Delete a project](#delete-a-project)
  - [Delete a record](#delete-a-record)
  - [Generate a report `[beta]`](#generate-a-report-beta)
  - [Print version information](#print-version-information)
- [Configuration](#configuration)
  - [Prefer 12-hour clock for storing records](#prefer-12-hour-clock-for-storing-records)
  - [Prefer decimal hours for status and reports](#prefer-decimal-hours-for-status-and-reports)
  - [Set your preferred editor](#set-your-preferred-editor)
  - [Configure defaults for projects](#configure-defaults-for-projects)
- [Credits](#credits)

---

## Installation

### Homebrew

```
brew tap dominikbraun/timetrace
brew install timetrace
```

### Snap

```
sudo snap install timetrace --edge --devmode
```

### AUR

```
yay -S timetrace-bin
```

### Scoop

```
scoop bucket add https://github.com/Br1ght0ne/scoop-bucket
scoop install timetrace
```

### Docker

The timetrace Docker image stores all data in the `/data` directory. To persist
this data on disk, you should create a bind mount or named volume like so:

```
docker container run -v my-volume:/data dominikbraun/timetrace version
```

### Binary

Download the [latest release](https://github.com/dominikbraun/timetrace/releases)
and extract the binary into a directory like `/usr/local/bin` or
`C:\Program Files\timetrace`. Make sure the directory is in the `PATH` variable.

## Usage example

First, create a project you're working for:

```
timetrace create project make-coffee
```

Once the project is created, you're able to track work on that project.

```
timetrace start make-coffee
```

You can obtain your currently worked time using `timetrace status`. When you've
finished your work, stop tracking:

```
timetrace stop
```

### Project modules

To refine what part of a project you're working on, timetrace supports _project modules_. These are the exact same thing
as normal projects, except that they have a key in the form `<module>@<project>`.

Creating a `grind-beans` module for the `make-coffee` project is simple:

```
timetrace create project grind-beans@make-coffee
```

The new module will be listed as part of the `make-coffee` project:

```
timetrace list projects
+-----+-------------+-------------+
|  #  |     KEY     |   MODULES   |
+-----+-------------+-------------+
|   1 | make-coffee | grind-beans |
+-----+-------------+-------------+

```

When filtering by projects, for example with `timetrace list records -p make-coffee today`, the modules of that project
will be included.

## Shell integration

### Starship

To integrate timetrace into Starship, add the following lines to `$HOME/.config/starship.toml`:

```
[custom.timetrace]
command = """ timetrace status --format "Current project: {project} - Worked today: {trackedTimeToday}" """
when = "timetrace status"
shell = "sh"
```

You can find a list of available formatting variables in the [`status` reference](#print-the-tracking-status).

## Command reference

### Start tracking

**Syntax:**

```
timetrace start <PROJECT KEY> [+TAG1, +TAG2, ...]
```

**Arguments:**

| Argument            | Description                                  |
| ------------------- | -------------------------------------------- |
| `PROJECT KEY`       | The key of the project.                      |
| `+TAG1, +TAG2, ...` | One or more optional tags starting with `+`. |

**Flags:**

| Flag             | Short | Description                                                                                                |
| ---------------- | ----- | ---------------------------------------------------------------------------------------------------------- |
| `--billable`     | `-b`  | Mark the record as billable.                                                                               |
| `--non-billable` |       | Mark the record as non-billable, even if the project is [billable by default](#per-project-configuration). |

**Example:**

Start working on a project called `make-coffee` and mark it as billable:

```
timetrace start --billable make-coffee
```

Start working on the `make-coffee` project and add two tags:

```
timetrace start make-coffee +espresso +morning
```

### Print the tracking status

**Syntax:**

```
timetrace status
```

**Flags:**

| Flag       | Short | Description                                                   |
| ---------- | ----- | ------------------------------------------------------------- |
| `--format` | `-f`  | Display the status in a custom format (see below).            |
| `--output` | `-o`  | Display the status in a specific output. Valid values: `json` |

**Formatting variables:**

The names of the formatting variables are the same as the JSON keys printed by `--output json`.

| Variable               | Description                              |
| ---------------------- | ---------------------------------------- |
| `{project}`            | The key of the current project.          |
| `{trackedTimeCurrent}` | The time tracked for the current record. |
| `{trackedTimeToday}`   | The time tracked today.                  |
| `{breakTimeToday}`     | The break time since the first record.   |

**Example:**

Print the current tracking status:

```
timetrace status
+-------------------+----------------------+----------------+
|  CURRENT PROJECT  |  WORKED SINCE START  |  WORKED TODAY  |
+-------------------+----------------------+----------------+
| make-coffee       | 1h 15min             | 4h 30min       |
+-------------------+----------------------+----------------+
```

Print the current project and the total working time as a custom string. Given the example above, the output will be
`Current project: make-coffee - Worked today: 3h 30min`.

```
timetrace status --format "Current project: {project} - Worked today: {trackedTimeToday}"
```

Print the status as JSON:

```
timetrace status -o json
```

The output will look as follows:

```json
{
  "project": "web-store",
  "trackedTimeCurrent": "1h 45min",
  "trackedTimeToday": "7h 30min",
  "breakTimeToday": "0h 30min"
}
```

### Stop tracking

**Syntax:**

```
timetrace stop
```

**Example:**

Stop working on your current project:

```
timetrace stop
```

### Create a project

**Syntax:**

```
timetrace create project <KEY>
```

**Arguments:**

| Argument | Description            |
| -------- | ---------------------- |
| `KEY`    | An unique project key. |

**Example:**

Create a project called `make-coffee`:

```
timetrace create project make-coffee
```

### Create a record

:warning: You shouldn't use this command for normal tracking but only for belated records.

**Syntax:**

```
timetrace create record <PROJECT KEY> {<YYYY-MM-DD>|today|yesterday} <HH:MM> <HH:MM>
```

**Arguments:**

| Argument      | Description                                                                      |
| ------------- | -------------------------------------------------------------------------------- |
| `PROJECT KEY` | The project key the record should be created for.                                |
| `YYYY-MM-DD`  | The date the record should be created for. Alternatively `today` or `yesterday`. |
| `HH:MM`       | The start time of the record.                                                    |
| `HH:MM`       | The end time of the record.                                                      |

**Example:**

Create a record for the `make-coffee` project today from 07:00 to 08:30:

```
timetrace create record make-coffee today 07:00 08:30
```

### Get a project

**Syntax:**

```
timetrace get project <KEY>
```

**Arguments:**

| Argument | Description      |
| -------- | ---------------- |
| `KEY`    | The project key. |

**Example:**

Display a project called `make-coffee`:

```
timetrace get project make-coffee
```

### Get a record

**Syntax:**

```
timetrace get record <YYYY-MM-DD-HH-MM>
```

**Arguments:**

| Argument           | Description                           |
| ------------------ | ------------------------------------- |
| `YYYY-MM-DD-HH-MM` | The start time of the desired record. |

**Example:**

By default, records can be accessed using the 24-hour format, meaning 3:00 PM is 15. Display a record created on May 1st 2021, 3:00 PM:

```
timetrace get record 2021-05-01-15-00
```

This behavior [can be changed](#prefer-12-hour-clock-for-storing-records).

### List all projects

**Syntax:**

```
timetrace list projects
```

**Example:**

List all projects stored within the timetrace filesystem:

```
timetrace list projects
+---+-------------+
| # |     KEY     |
+---+-------------+
| 1 | make-coffee |
| 2 | my-website  |
| 3 | web-shop    |
+---+-------------+
```

### List all records from a date

**Syntax:**

```
timetrace list records {<YYYY-MM-DD>|today|yesterday}
```

**Arguments:**

| Argument     | Description                                                 |
| ------------ | ----------------------------------------------------------- |
| `YYYY-MM-DD` | The date of the records to list, or `today` or `yesterday`. |
| today        | List today's records.                                       |
| yesterday    | List yesterday's records.                                   |

**Flags:**

| Flag         | Short | Description                    |
| ------------ | ----- | ------------------------------ |
| `--billable` | `-b`  | only display billable records. |
| `--project`  | `-p`  | filter records by project key. |

**Example:**

Display all records created on May 1st 2021:

```
timetrace list records 2021-05-01
+-----+-------------+---------+-------+------------+
|  #  |   PROJECT   |  START  |  END  |  BILLABLE  |
+-----+-------------+---------+-------+------------+
|   1 | my-website  | 17:30   | 21:00 | yes        |
|   2 | my-website  | 08:31   | 17:00 | no         |
|   3 | make-coffee | 08:25   | 08:30 | no         |
+-----+-------------+---------+-------+------------+
```

Filter records by the `make-coffee` project:

```
timetrace list records -p make-coffee 2021-05-01
+-----+-------------+---------+-------+------------+
|  #  |   PROJECT   |  START  |  END  |  BILLABLE  |
+-----+-------------+---------+-------+------------+
|   1 | make-coffee | 08:25   | 08:30 | no         |
+-----+-------------+---------+-------+------------+
```

This will include records for [project modules](#project-modules) like `grind-beans@make-coffee`.

### Edit a project

**Syntax:**

```
timetrace edit project <KEY>
```

**Arguments:**

| Argument | Description      |
| -------- | ---------------- |
| `KEY`    | The project key. |

**Flags:**
| Flag       | Short | Description                                             |
| ---------- | ----- | ------------------------------------------------------- |
| `--revert` | `-r`  | Revert the project to its state prior to the last edit. |

**Example:**

Edit a project called `make-coffee`:

```
timetrace edit project make-coffee
```

:fire: **New:** Restore the project to its state prior to the last edit:

```
timetrace edit project make-coffee --revert
```

### Edit a record

**Syntax:**

```
timetrace edit record {<KEY>|latest}
```

**Arguments:**

| Argument | Description                                                                                                                                 |
| -------- | ------------------------------------------------------------------------------------------------------------------------------------------- |
| `KEY`    | The project key. `YYYY-MM-DD-HH-MM` by default or `YYYY-MM-DD-HH-MMPM` if [`use12hours` is set](#prefer-12-hour-clock-for-storing-records). |

**Flags:**

| Flag       | Short | Description                                                                   |
| ---------- | ----- | ----------------------------------------------------------------------------- |
| `--plus`   | `-p`  | Add the given duration to the record's end time, e.g. `--plus 1h 10m`         |
| `--minus`  | `-m`  | Subtract the given duration from the record's end time, e.g. `--minus 1h 10m` |
| `--revert` | `-r`  | Revert the record to its state prior to the last edit.                        |

**Example:**

Edit the latest record. Specifying no flag will open the record in your editor:

```
timetrace edit record latest
```

Add 15 minutes to the end of the record created on May 1st, 3PM:

```
timetrace edit record 2021-05-01-15-00 --plus 15m
```

:fire: **New:** Restore the record to its state prior to the last edit:

```
timetrace edit record 2021-05-01-15-00 --revert
```

Tip: You can get the record key `2021-05-01-15-00` using [`timetrace list records`](#list-all-records-from-a-date).

### Delete a project

**Syntax:**

```
timetrace delete project <KEY>
```

**Arguments:**

| Argument | Description      |
| -------- | ---------------- |
| `KEY`    | The project key. |

**Flags:**

| Flag                | Short | Description                                                                                                                             |
| ------------------- | ----- | --------------------------------------------------------------------------------------------------------------------------------------- |
| `--revert`          | `-r`  | Restore a deleted project.                                                                                                              |
| `--exclude-records` | `-e`  | Exclude associated project records from the deletion. If used together with `--revert`, excludes restoring project records from backup. |

**Example:**

Delete a project called `make-coffee`. Note that submodules will be deleted along with the parent project:

```
timetrace delete project make-coffee
```
The command will prompt for confirmation of whether project records should be deleted too.

:fire: **New:** Restore the project to its pre-deletion state. Submodules will be restored along with the parent project:

```
timetrace delete project make-coffee --revert
```
The command will prompt for confirmation of whether project records should be restored from backup too. This is a
potentially dangerous operation since records edited in the meantime will be overwritten by the backup.

### Delete a record

**Syntax:**

```
timetrace delete record <YYYY-MM-DD-HH-MM>
```

**Arguments:**

| Argument           | Description                           |
| ------------------ | ------------------------------------- |
| `YYYY-MM-DD-HH-MM` | The start time of the desired record. |

| Flag       | Short | Description                 |
| ---------- | ----- | --------------------------- |
| `--yes`    |       | Do not ask for confirmation |
| `--revert` | `-r`  | Restore a deleted record.   |

**Example:**

Delete a record created on May 1st 2021, 3:00 PM:

```
timetrace delete record 2021-05-01-15-00
```

:fire: **New:** Restore the record to its pre-deletion state:

```
timetrace delete record 2021-05-01-15-00 --revert
```

### Generate a report `[beta]`

**Syntax:**

```
timetrace report
```

**Flags:**

| Flag                    | Short | Description                                                                                                                                                        |
| ----------------------- | ----- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| `--billable`            | `-b`  | Filter report for only billable records.                                                                                                                           |
| `--non-billable`        |       | Filter report for non-billable records.                                                                                                                            |
| `--start <YYYY-MM-DD>`  | `-s`  | Filter report from a specific point in time (start is inclusive).                                                                                                  |
| `--end <YYYY-MM-DD>`    | `-e`  | Filter report to a specific point in time (end is inclusive).                                                                                                      |
| `--project <KEY>`       | `-p`  | Filter report for only one project.                                                                                                                                |
| `--output <json>`       | `-o`  | Write report as JSON to file.                                                                                                                                      |
| `--file path/to/report` | `-f`  | Write report to a specific file <br>(if not given will use config `report-dir`<br> if config not present writes to `$HOME/.timetrace/reports/report-<time.unix>`). |

### Print version information

**Syntax:**

```
timetrace version
```

**Example:**

Print your installed timetrace version:

```
timetrace version
```

## Configuration

You may provide your own configuration in a file called `config.yaml` within
`$HOME/.timetrace`.

### Prefer 12-hour clock for storing records

If you prefer to use the 12-hour clock instead of the default 24-hour format,
add this to your `config.yaml` file:

```yaml
# config.yml
use12hours: true
```

This will allow you to [view a record](#get-a-record) created at 3:00 PM as
follows:

```
timetrace get record 2021-05-14-03-00PM
```

### Prefer decimal hours for status and reports

If your prefer to use decimal hours for durations, e.g. `1.5h` instead of `1h 30m`,
add this to your `config.yaml` file:

```yaml
useDecimalHours: "On"
```

To display durations in _both_ formats at the same time, use:

```yaml
useDecimalHours: "Both"
```

**Examples with durations in different formats:**

```
default (useDecimalHours = "Off")
+-------------------+----------------------+----------------+----------+
|  CURRENT PROJECT  |  WORKED SINCE START  |  WORKED TODAY  |  BREAKS  |
+-------------------+----------------------+----------------+----------+
| make-coffee       | 1h 8min              | 3h 8min        | 0h 11min |
+-------------------+----------------------+----------------+----------+

Decimal Hours (useDecimalHours = "On")
+-------------------+----------------------+----------------+----------+
|  CURRENT PROJECT  |  WORKED SINCE START  |  WORKED TODAY  |  BREAKS  |
+-------------------+----------------------+----------------+----------+
| make-coffee       | 1.2h                 | 3.2h           | 0.2h     |
+-------------------+----------------------+----------------+----------+

Both (useDecimalHours = "Both")
+-------------------+----------------------+----------------+---------------+
|  CURRENT PROJECT  |  WORKED SINCE START  |  WORKED TODAY  |    BREAKS     |
+-------------------+----------------------+----------------+---------------+
| make-coffee       | 1h 8min 1.2h         | 3h 8min 3.2h   | 0h 11min 0.2h |
+-------------------+----------------------+----------------+---------------+
```

### Set your preferred editor

By default, timetrace will open the editor specified in `$EDITOR` or fall back
to `vi`. You may set your provide your preferred editor like so:

```yaml
# config.yml
editor: nano
```

### Configure defaults for projects

To add a configuration for a specific project, use the `projects` key which accepts
a map with the project key as key and the project configuration as value.

Each project configuration currently has the following schema:

```yaml
billable: bool
```

For example, always make records for the `make-coffee` project billable:

```yaml
# config.yml
projects:
    make-coffee:
        billable: true
```

## Credits

This project depends on the following packages:

- [spf13/cobra](https://github.com/spf13/cobra)
- [spf13/viper](https://github.com/spf13/viper)
- [fatih/color](https://github.com/fatih/color)
- [olekukonko/tablewriter](https://github.com/olekukonko/tablewriter)
- [enescakir/emoji](https://github.com/enescakir/emoji)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.14.3] - 2022-03-06

### Fixed
* Fix output for hour format (#208)

## [0.14.2] - 2021-10-27

### Changed
* Make `tzdata` available in timetrace Docker image

## [0.14.1] - 2021-10-22

### Fixed
* Fixed panic when providing no argument for `timetrace start` (#199)

## [0.14.0] - 2021-10-02

### Added
* Add support for tags (#183)
* Add support for durations in decimal format (#180)
* Introduce `@ID` notation for selecting records (#116)

### Changed
* Print colliding records as table when a collision occurs (#174)

## [0.13.0] - 2021-08-10

### Added
* Add support for per-project configuration (#166)
* Add support for making projects billable by default (#166)
* Add support for deleting and reverting projects with their modules (#160)
* Add support for restoring records associated with a restored project (#160)
* Add `--non-billable` flag for `timetrace start` command (#172) 
* Add `--non-billable` flag for `timetrace report` command (#142)

### Changed
* Ask the user whether they want to delete all records when deleting a project (#160)

## [0.12.0] - 2021-07-30

### Changed
* `timetrace report`: Include project modules when filtering for a project (#143)

### Fixed
* Don't consider any backup records for calculations (#162)

### Other
* Refactorings and optimizations (#146, #164)
* Preparations for the larger and relevant 0.13.0 release

## [0.11.1] - 2021-07-05

### Changed
* Default to 'no' when asking the user for confirmation (#147)
* Ask the user for confirmation when deleting a project (#147, #155)
* Don't allow creation of records in the future (#152)

### Fixed
* Fix nil pointer dereference when checking record collisions (#150)
* Fix typos in documentation (#149)

## [0.11.0] - 2021-06-26

### Added
* Add `create record` command (#118)
* Add beta support for reports (#99)

## [0.10.0] - 2021-06-15

### Added
* Add `--output` flag for `timetrace status` (#129)
* Add `--format` flag for `timetrace status` (#114, #123, #124)
* Add Starship support
* Add Scoop support

### Changed
* Remove seconds from printed durations (#113)

## [0.9.0] - 2021-06-09

### Added
* Add the total tracked time for all listed records to `timetrace list records` (#106)

## [0.8.0] - 2021-06-06

### Added
* Introduce the `--revert` flag for `edit record`, `delete record`, `edit project` and `delete project` (#93)
* Add the overall break time for `timetrace status` (#100)

## [0.7.2] - 2021-06-04

### Fixed
* Fix critical error when starting tracking if there are no existing records (#103)

## [0.7.1] - 2021-06-02

### Fixed
* Fix unhandled error in `edit record` command if there are no tracked records (#96)

## [0.7.0] - 2021-05-30

### Added
* Add support for project modules

### Changed
* Consider project modules when filtering projects (#63)
* Display project modules when listing projects (#70)
* Require parent projects to exist when creating a module (#80)

## [0.6.1] - 2021-05-26

### Fixed
* Fix `timetrace stop` command (#86)

## [0.6.0] - 2021-05-26

### Added
* Add `timetrace edit record` command (#51)
* Add `latest` alias for `timetrace edit record` (#73)

### Changed
* Use an info output for `timetrace status` when there's no active tracking (#65)
* Always adhere to the `use12hours` setting for date- and time input and output (#67)

### Fixed
* Don't allow edting of incomplete records (#69)
* Don't allow re-creation of existing projects (#79)

## [0.5.0] - 2021-05-22

### Added
* Add record key to `list records` output (#44)
* Add `--project` filter to `list records` (#62)

### Changed
* Make command syntax and help sensible to `use12hours` in `config.yaml` (#44, #54)

### Fixed
* Fix Docker image labels (#35)

## [0.4.0] - 2021-05-20

### Added
* Add support for Bash autocompletion (#25)
* Add support for Snap (#31)
* Add filter for billable records (#33)

### Changed
* Use non-error output for `timetrace version` (#45)
* Colorize and stylize tables (#26, #48, #49)

## [0.3.1] - 2021-05-18

### Changed
* Introduce the `today` and `yesterday` aliases for `timetrace list records` (#32)
* Preview the target record and ask for confirmation when deleting a record (#37)

## [0.3.0] - 2021-05-17

### Added
* `timetrace delete record` command (#22)

### Changed
* Add project name to the output of `list records` (#24)

## [0.2.1] - 2021-05-16

### Changed
* Beautified output of time durations (#20)

### Fixed
* Fix `timetrace status` if there was no time tracking yet (#18)

## [0.2.0] - 2021-05-16

### Added
* `timetrace list records` command (#16)

## [0.1.4] - 2021-05-14

### Fixed
* Fix calculation of total tracked time

## [0.1.3] - 2021-05-14

### Fixed
* Fix determination of latest stored record

## [0.1.2] - 2021-05-14

### Fixed
* Support slashes and back-slashes in project keys (#9)

## [0.1.1] - 2021-05-14

### Fixed
* Fix status report for situations where time tracking is not active (#7)

## [0.1.0] - 2021-05-13

### The initial timetrace release.


fs/fs.go
package fs

import (
	"fmt"
	"io/ioutil"
	"os"
	"path"
	"path/filepath"
	"sort"
	"strconv"
	"strings"
	"time"

	"github.com/dominikbraun/timetrace/config"
	"github.com/spf13/afero"
)

const (
	rootDirName     = ".timetrace"
	projectsDirName = "projects"
	recordsDirName  = "records"
	reportDirName   = "reports"
)

const (
	recordDirLayout            = "2006-01-02"
	recordFilepathLayout       = "15-04.json"
	recordBackupFilepathLayout = "15-04.json.bak"
)

type Fs struct {
	Wrapper   afero.Fs
	config    *config.Config
	sanitizer *strings.Replacer
}

func New(config *config.Config) *Fs {
	return &Fs{
		Wrapper:   afero.NewMemMapFs(), // change later
		config:    config,
		sanitizer: strings.NewReplacer("/", "-", "\\", "-"),
	}
}

// ProjectFilepath returns the filepath of the project with the given key.
func (fs *Fs) ProjectFilepath(key string) string {
	key = fs.sanitizer.Replace(key)
	name := fmt.Sprintf("%s.json", key)
	return filepath.Join(fs.projectsDir(), name)
}

// ProjectBackupFilepath return the filepath of the backup project with the
// given key.
func (fs *Fs) ProjectBackupFilepath(key string) string {
	key = fs.sanitizer.Replace(key)
	name := fmt.Sprintf("%s.json.bak", key)
	return filepath.Join(fs.projectsDir(), name)
}

// ProjectFilepaths returns all non-backup project filepaths sorted alphabetically.
func (fs *Fs) ProjectFilepaths() ([]string, error) {
	dir := fs.projectsDir()

	items, err := afero.ReadDir(fs.Wrapper, dir)
	if err != nil {
		return nil, err
	}

	var filepaths []string

	for _, item := range items {
		if item.IsDir() {
			continue
		}
		itemName := item.Name()
		if strings.HasSuffix(itemName, ".bak") {
			continue
		}

		filepaths = append(filepaths, filepath.Join(dir, itemName))
	}
	sort.Strings(filepaths)

	return filepaths, nil
}

// ProjectBackupFilepaths returns all backup project filepaths sorted alphabetically.
func (fs *Fs) ProjectBackupFilepaths() ([]string, error) {
	dir := fs.projectsDir()

	items, err := ioutil.ReadDir(dir)
	if err != nil {
		return nil, err
	}

	var filepaths []string

	for _, item := range items {
		if item.IsDir() {
			continue
		}
		itemName := item.Name()
		if !strings.HasSuffix(itemName, ".bak") {
			continue
		}

		filepaths = append(filepaths, filepath.Join(dir, itemName))
	}
	sort.Strings(filepaths)

	return filepaths, nil
}

// RecordFilepath returns the filepath of the record with the given start time.
//
// Note that the start time also has to contain the date as this determines the
// directory the project is stored in.
func (fs *Fs) RecordFilepath(start time.Time) string {
	name := start.Format(recordFilepathLayout)
	return filepath.Join(fs.RecordDirFromDate(start), name)
}

func (fs *Fs) RecordBackupFilepath(start time.Time) string {
	name := start.Format(recordBackupFilepathLayout)
	return filepath.Join(fs.RecordDirFromDate(start), name)
}

// RecordFilepaths returns all non-backup record filepaths within the given
// directory sorted by the given function.
//
// The directory can be obtained using functions like recordDir or RecordDirs.
// If you have a record date, use RecordDirFromDate to get the directory name.
//
// The less function allows you to sort the records. Assume three record files:
//
//	- timetrace/records/2021-05-01/08-00.json
//	- timetrace/records/2021-05-01/10-00.json
//	- timetrace/records/2021-05-01/11-30.json
//
// The following call to RecordFilepaths will return the paths of those records
// sorted from newest to oldest:
//
//	latestRecords, err := RecordFilepaths(dir, func (a, b string) bool {
//		timeA, _ := time.Parse("15-04.json", a)
//		timeB, _ := time.Parse("15-04.json", b)
//		return timeA.Before(timeB)
//	})
//
// This can be used to determine the latest record in a given record directory.
func (fs *Fs) RecordFilepaths(dir string, less func(a, b string) bool) ([]string, error) {
	items, err := ioutil.ReadDir(dir)
	if err != nil {
		return nil, err
	}

	filepaths := make([]string, 0)

	for _, item := range items {
		if item.IsDir() {
			continue
		}
		itemName := item.Name()
		if strings.HasSuffix(itemName, ".bak") {
			continue
		}

		filepaths = append(filepaths, filepath.Join(dir, itemName))
	}

	sort.Slice(filepaths, func(i, j int) bool {
		return less(filepaths[i], filepaths[j])
	})

	return filepaths, nil
}

// RecordDirs returns all record directories sorted alphabetically. This can be
// used to determine the latest record directory and obtain the latest record
// within that directory using RecordFilepaths.
//
// Note that all timetrace directories have to exist for RecordDirs to work.
func (fs *Fs) RecordDirs() ([]string, error) {
	items, err := ioutil.ReadDir(fs.recordsDir())
	if err != nil {
		return nil, err
	}

	var dirs []string

	for _, item := range items {
		if !item.IsDir() {
			continue
		}
		dirs = append(dirs, filepath.Join(fs.recordsDir(), item.Name()))
	}

	sort.Strings(dirs)

	return dirs, nil
}

func (fs *Fs) RecordDirFromDate(date time.Time) string {
	dir := date.Format(recordDirLayout)
	return fs.recordDir(dir)
}

// EnsureDirectories creates all required timetrace directories. If they already
// exist, nothing happens.
func (fs *Fs) EnsureDirectories() error {
	dirs := []string{
		fs.projectsDir(),
		fs.recordsDir(),
		fs.recordsInitSubDir(),
		fs.ReportDir(),
	}

	for _, dir := range dirs {
		if err := fs.Wrapper.MkdirAll(dir, 0777); err != nil {
			return err
		}
	}

	return nil
}

func (fs *Fs) EnsureRecordDir(date time.Time) error {
	return os.MkdirAll(fs.RecordDirFromDate(date), 0777)
}

func (fs *Fs) recordDir(name string) string {
	return filepath.Join(fs.recordsDir(), name)
}

func (fs *Fs) projectsDir() string {
	return filepath.Join(fs.rootDir(), projectsDirName)
}

func (fs *Fs) recordsDir() string {
	return filepath.Join(fs.rootDir(), recordsDirName)
}

func (fs *Fs) recordsInitSubDir() string {
	return fs.RecordDirFromDate(time.Now())
}

func (fs *Fs) ReportDir() string {
	return path.Join(fs.rootDir(), reportDirName)
}

func (fs *Fs) rootDir() string {
	if fs.config.Store != "" {
		return os.ExpandEnv(fs.config.Store)
	}

	homeDir, _ := os.UserHomeDir()

	return filepath.Join(homeDir, rootDirName)
}

func (fs *Fs) WriteReport(filepath string, data []byte) error {
	reportPath := filepath
	// no out flag provided
	// TODO: looks a little irritating could be re-written
	if reportPath == "" {
		reportPath = fs.config.ReportPath
		if reportPath == "" {
			// fileName -> report-<time.unix>
			fileName := strings.Join([]string{"report", strconv.FormatInt(time.Now().Unix(), 10)}, "-")
			reportPath = path.Join(fs.ReportDir(), fileName)
		}
	}

	if err := ioutil.WriteFile(reportPath, data, 0644); err != nil {
		return err
	}
	return nil
}

func (fs *Fs) Stat(path string) (*os.FileInfo, error) {
	return nil, fmt.Errorf("not implemented")
}
